### DISCUSSION TITLE
Weird crash on iOS 15

### DISCUSSION THREAD

**Discussion by buscarini** [pointfreeco/swift-composable-architecture#986] [Q&A] [answered]
I'm seeing this crash when running on a device with iOS 15. It seems more like a bug with the code generated by the compiler, but I haven't been able to find much info on the Internet. Has anyone encountered this? Any workaround? I have tried to disable bitcode, optimizations, but it crashes anyway.

<img width="2227" alt="Captura de pantalla 2022-01-27 a las 8 51 53" src="https://user-images.githubusercontent.com/486734/151315351-eecc59a3-a82d-4989-9a76-b4f0b2462073.png">

**Comment by tgrapperon** [comment, marked as answer]
In my experience, these weird crashes (that are happening where they shouldn't) are caused by a stack overflow. It can happen easily when using TCA, as the State can end up being a huge struct value that lives almost entirely on the stack.
Maybe you can try to box a few of your child states (you can convert them as a computed property to the first value of a one-value array (which stores its values on the heap) to quickly check if it helps, or use a proper boxing property wrapper). You may find more information [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/488) and [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/752) for example.

It may be something else, but that's what I'd try first in this case. Deleting DerivedData may also be an option if you have doubts about your compiler-generated code.

**Reply by tgrapperon** [reply]
BTW, looking at the StackTrace again, and I'm maybe reading it wrong, but it looks like you're observing some quite complex data (`AppState>TabState>HomeModel>SelectionState`) through a `ViewStore`. If this is indeed the case, this is not optimal performance-wise. As it is unlikely that your UI will need access to all the `AppState` at this level, you should maybe consider scoping your store to an intermediate `ViewState` that exposes only the bare minimum to construct the UI.

If this is not a stack overflow, the issue seems related to the equatability of `SelectionState`. Maybe you can build two `SelectionState` values at launch and see if it crashes too when you compare them.

But as @jaanussiim said, it is faster to clean the build folder at first, as it requires no code change. If this doesn't solve the issue, and `SelectionState` is failing to compare in isolation, the next level is deleting DerivedData. If you do this, you should quit Xcode first. You will also need to "Xcode>Package>Reset Package Cache" when you launch it back.

**Comment by jaanussiim** [comment]
I usually see this when changing the state object in module/package between runs. Quick 'Clean Build Folder' resolves it. 

I would blame Xcode. May be related to the new experimental build system in Xcode 13.2. 

Working with smaller feature modules helps to reduce recompiling needed.

**Comment by buscarini** [comment]
Thanks all. Yes, I saw that there was a view store inspecting the whole app state, and optimized that away. I got rid of some code and this crash went away. Probably it's due to a stack overflow, as @tgrapperon suggested; I discarded it initially because the stack didn't seem that large to me, but it might be related.
